AP-1092 | 2 | Added error handling for fetching follows

vershwal opened the pull request:
Ref https://linear.app/ghost/issue/AP-1092

- Used PersistedAccount type for siteDefaultAccount - the current site account cannot be null.
- Used `Result` pattern for fetching follows
- Used `apId` instead of `handle` - same as we have for fetching posts
- If fetching any one follow fails, skipping that so that we can still return the other valid follows in the collection
- Removed getFollowsByApId, the controller now decides if we want to do a local lookup or a remote lookup.
- Modified tests to test getFollowsByAccount and getFollowsByRemoteLookUp separately

mike182uk commented on src/http/api/views/account.follows.view.ts:
```diff
@@ -183,17 +183,22 @@ export class AccountFollowsView {
-                const actor = (await item.toJsonLd({
+        const iterator = page.getItems()[Symbol.asyncIterator]();
+
+        while (true) {
```
Why use `while(true)` over iterating over `AsyncIterable`?

vershwal replied:
The problem I am trying to solve here is that - `page.getItems()` fails for some items and because it's a part of for loop, there is no error handling so this returns 500. But yeah, while(true) wasn't a good way so I changed the approach a bit.  Used `page.itemIds` this will only get the url and won't fail even if the account is not present in activitypub servers, now we fetch the actual object in a try-catch that will handle the failures gracefully.

vershwal commented on src/http/api/views/account.follows.view.ts:
```diff
@@ -68,7 +77,7 @@ export class AccountFollowsView {
-        siteDefaultAccount: Account,
+        siteDefaultAccount: PersistedAccount,
     ): Promise<AccountFollows> {
         if (!siteDefaultAccount.id) {
```
You are right, removed it!

vershwal commented on src/http/api/views/account.follows.view.ts:
```diff
@@ -218,20 +223,22 @@ export class AccountFollowsView {
-        return {
+        return ok({
             accounts: accounts,
             total: 0,
```
The field `total` is not actually used on the UI. Code [here](https://github.com/TryGhost/Ghost/blob/main/apps/admin-x-activitypub/src/api/activitypub.ts#L69)
It was redundant. I cleaned it up.

vershwal commented on src/http/api/views/account.follows.view.ts:
```diff
@@ -218,20 +223,22 @@ export class AccountFollowsView {
+            } catch {
+                // Skip this item if processing fails
+                // This ensures that a single invalid or unreachable follow doesn't block the API from returning valid follows
+                // If fetching any one follow fails, we can still return the other valid follows in the collection
```
Done

vershwal commented on src/http/api/account.ts:
```diff
@@ -136,13 +139,47 @@ export function createGetAccountFollowsHandler(
             );
+            if (isError(result)) {
+                const error = getError(result);
+                switch (error) {
```
Added tests!

vershwal commented on src/http/api/views/account.follows.view.ts:
```diff
@@ -38,26 +45,28 @@ export class AccountFollowsView {
-        account: Account | null,
+    async getFollowsByApId(
+        apId: URL,
+        account: Account,
```
I messed this while resolving merge commits. Fixed it.

vershwal replied:
I messed this up while resolving merge commits. Fixed it.

allouis commented on src/http/api/account.ts:
```diff
@@ -136,20 +139,51 @@ export function createGetAccountFollowsHandler(
 
-            accountFollows = await accountFollowsView.getFollowsByHandle(
-                handle,
+            const result = await accountFollowsView.getFollowsByApId(
```
Should we be calling `getFollowsByAccount` if `account` is present?

allouis replied:
Or we move the local existence check into the `getFollowsByApId` method 🤔 But right now the knowledge is kinda split between the two

allouis reviewed (changes requested):
We're not actually testing `getFollowsByApId` here, and we maybe have a bug where we're passing `offset` instead of `next`? But i'm not 100% clear

I think either `getFollowsByApId` contains the logic to check for a local account - and the tests are testing this method OR the controller determine whether to call the remote or local lookup based on whether we get an `Account` back from the repository, and the `getFollowsByApId` method is deleted

vershwal commented:
- Removed `getFollowsByApId`.
- The controller now decides if we want to do a local lookup or a remote lookup. 
- Modified tests to test `getFollowsByAccount` and `getFollowsByRemoteLookUp` separately 

> We are passing offset here instead of next
- This was just bad naming, we are actually sending next. Fixed the naming.
